src/CoreWCF.Primitives/tests/ServiceKnownTypeTests.cs

```diff
@@ -0,0 +1,223 @@
+            TContract channel = factory.CreateChannel();
+            ((System.ServiceModel.Channels.IChannel)channel).Open();
+            return channel;
+        }
+
+        internal static void CloseChannel<TContract>(System.ServiceModel.ChannelFactory<TContract> factory, TContract channel)
+        {
+            ((System.ServiceModel.Channels.IChannel)channel).Close();
+            factory.Close();
+            TestHelper.CloseServiceModelObjects((System.ServiceModel.Channels.IChannel)channel, factory);
+        }
+
+        [Fact]
+        public static void ServiceKnownTypeException()
+        {
+            var factory = ServiceKnownTypeTests.CreateFactory<IPingServiceEx, PingServiceEx>();
+            var channel = ServiceKnownTypeTests.CreateChannel(factory);
+            Assert.Throws<System.Xml.XmlException>(() =>
+            {
+                BaseMsg msg = channel.Ping("hello");
+            });
+            ServiceKnownTypeTests.CloseChannel(factory, channel);
+        }
+
+        [Fact]
+        public static void ServiceKnownTypeTest1()
+        {
+            var factory = ServiceKnownTypeTests.CreateFactory<IPingService1, PingService1>();
+            var channel = ServiceKnownTypeTests.CreateChannel(factory);
+            BaseMsg msg = channel.Ping("hello");
+            Assert.Equal("hello", msg.msg);
+            ServiceKnownTypeTests.CloseChannel(factory, channel);
+        }
+
+        [Fact]
+        public static void ServiceKnownTypeTest2()
+        {
+            var factory = ServiceKnownTypeTests.CreateFactory<IPingService2, PingService2>();
+            var channel = ServiceKnownTypeTests.CreateChannel(factory);
+            BaseMsg msg = channel.Ping("hello");
+            Assert.Equal("hello", msg.msg);
+            ServiceKnownTypeTests.CloseChannel(factory, channel);
+        }
+
+        [Fact]
+        public static void ServiceKnownTypeTest3()
+        {
+            var factory = ServiceKnownTypeTests.CreateFactory<IPingService3, PingService3>();
+            var channel = ServiceKnownTypeTests.CreateChannel(factory);
+            BaseMsg msg = channel.Ping("hello");
+            Assert.Equal("hello", msg.msg);
+            ServiceKnownTypeTests.CloseChannel(factory, channel);
+        }
+
+        [Fact]
+        public static void ServiceKnownTypeTest4()
+        {
+            var factory = ServiceKnownTypeTests.CreateFactory<IPingService4, PingService4>();
+            var channel = ServiceKnownTypeTests.CreateChannel(factory);
+            BaseMsg msg = channel.Ping("hello");
+            Assert.Equal("hello", msg.msg);
+            ServiceKnownTypeTests.CloseChannel(factory, channel);
+        }
+    }
+
+    internal class PingServiceEx : IPingServiceEx
+    {
+        public BaseMsg Ping(string msg)
+        {
+            return new DerivedMsg { msg = msg };
+        }
+
+    }
+
+    [ServiceContract]
+    [System.ServiceModel.ServiceContract]
+    public interface IPingServiceEx
```

Comment:
The set of attributes is a bit overwhelming to be able to quickly look over at a glance. We already have tests to validate `System.ServiceModel.ServiceContract` vs `CoreWCF.ServiceContract` (and the same for `OperationContract`) so we can assume you can just use `System.ServiceModel.ServiceContract` as the compatibility just works and leave out the CoreWCF namespaced one (same for OperationContract).

Reply:
Done